### ISSUE TITLE
✈️  Give to package.json

### ISSUE THREAD

**Issue by whit537** [gratipay/gratipay.com#4427]
-------------------------------

✈️  **This is the flight deck for the [Give to package.json](https://github.com/orgs/gratipay/projects/14?fullscreen=true) project.** ✈️ 

-------------------------------

🗻  **That project is part of the [Make it easier for companies to fund open source](https://github.com/gratipay/inside.gratipay.com/issues/987) epic.** 🗻 

-------------------------------

The definition of done is when we can say "Give us $10,000 and a package.json, and we'll distribute it for you."

### Todo

- [x] make room on the `/on/npm/foo/` page for the claiming workflow—move the name to the banner h1, and replace the wombat with the description: #4304
- [x] new email subflooring: ~~#4343~~ ~~#4344~~ ~~#4345~~ ~~#4365~~ ~~#4350~~ ~~#4355~~ ~~#4346~~ ~~#4347~~ ~~#4348~~ ~~#4349~~ ~~#4383~~
- [x] pull emails from maintainers only, not author (https://github.com/gratipay/gratipay.com/pull/4305#issuecomment-278429627): https://github.com/gratipay/gratipay.com/pull/4438
  - [x] also dedupe email addresses: https://github.com/gratipay/gratipay.com/pull/4438
- [x] set up npm syncing—besides just needing to stay in sync, people are going to want to ~~update the email address on file in their package.json~~ ~~release a new version~~ *take some action at npm* and then "immediately" see the new email show up to continue the flow: https://github.com/gratipay/gratipay.com/pull/4438
  - [x] add sync lag to [health dashboard](http://inside.gratipay.com/appendices/health)
- [x] Implement claim/unclaim/reclaim packages: #4305.
- [x] bulk add from `/on/npm/`: #4488
- [x] discovery via package.json: #4507
- [ ] fix bugs
- [x] write blog post justifying amount recommendation
- [x] "Pay for Open Source" on the homepage: #4598
- [ ] update payment processing howtos: https://github.com/gratipay/inside.gratipay.com/pull/1154
- [ ] write howto for managing mailing list of company payers
- [ ] Takeoff on the [project board](https://github.com/orgs/gratipay/projects/14?fullscreen=true)

-------------------------------

✈️  **This is the flight deck for the [Give to package.json](https://github.com/orgs/gratipay/projects/14?fullscreen=true) project.** ✈️ 

-------------------------------

**Comment by whit537** [comment]
https://github.com/gratipay/gratipay.com/pull/4305 is awesome ... but how does one get to a package page in the first place? Search? Browse? Edit URL? Eventually #4416 will be a nice way to get there based on email, but even then we'll still a way to get to a particular package based on its name as well.

**Comment by whit537** [comment]
Kicking out a bunch of new tickets in Takeoff. Need to stay focused on #4305.

**Comment by whit537** [comment]
Alright, #4305 is topped up, consolidated w/ three remaining pretty manageable PRs.

Looking beyond that, we need to get syncing going again. That's a pre-req and a not insignificant one, though we had some implementation of that before and there's code around somewhere towards it. I guess that's the primary focus next week.

Once syncing is out and we deploy #4305, there are a few directions we can go:

- **Cleanup**—There are lots of little inefficiencies piled up on the runway ready for takeoff. Which are highest value?
- **Discovery**—There is no way right now to get from gratipay.com to any given npm package page. Search? Browse? Bulk-add?
- **Outreach**—We can start emailing individual package maintainers inviting them to apply to join, with links to their specific packages.

**Comment by whit537** [comment]
Alright, scope clarified. This is about the "Give us $10,000 and a package.json" MVP. What's left for that is syncing, and then pledging (https://github.com/gratipay/gratipay.com/issues/4195), and then a UI for uploading a package.json like we had started looking at in #4135. How far can we get this week?

**Comment by whit537** [comment]
I'm reading through our existing npm syncing code and retracing why we turned it off (we turned it off, didn't we?) and thinking through how to turn it back on.

**Comment by whit537** [comment]
I'm finding an initial attempt at https://github.com/gratipay/gratipay.com/issues/4148#issuecomment-259559478 (same thread as our [tech debt discussion](https://github.com/gratipay/gratipay.com/issues/4148#issuecomment-264540868), btw).

**Comment by whit537** [comment]
Alright, the way I'm reading https://github.com/gratipay/gratipay.com/issues/4148 (and jogging my memory) is that the initial plan of running a worker dyno to continually update from npm did not work out because it was too buggy. So I [gave up](https://github.com/gratipay/gratipay.com/issues/4148#issuecomment-256541272) on running the syncer continuously and instead focused on getting it to run once for an initial load. We did fix a number of issues in the syncer after that point but never got back around to setting it up to run continuously. The first successful load took five days though so I think we should look into some better parallelization.

**Comment by whit537** [comment]
Moving to PR in https://github.com/gratipay/gratipay.com/pull/4438 ...

**Comment by whit537** [comment]
Alright, we're out of time to deploy npm claiming (#4305) this week. I [fly out](https://github.com/gratipay/inside.gratipay.com/issues/948) in 18 hours and then go straight into a weekend conference. It's too risky to try to deploy #4305 before then. I'd be fine to deploy npm syncing (#4438), though. I'm going to try to get that ready for review in the next few hours.

@dowski @rohitpaulk If either of you have any cycles in the next 12 hours to review that, that would be awesome. Then we can deploy tomorrow morning and let it run over the weekend. Should be pretty low risk since it'll be in a separate worker process operating on a barely-used table (`packages`), and it'll be good to have a couple days experience with how it behaves before deploying #4305 on top of it.

Let's aim for Monday for the big #4305 rollout. @clone1018 and I will be together in Austin and we will have Monday and Tuesday free. #4305 and #4441 are the two top priorities for those days.

This is happening, people!

!m * 💸  💸 💸 💸 💃 🌻

**Comment by whit537** [comment]
We are live with package claiming!

![screen shot 2017-05-09 at 7 27 43 pm](https://cloud.githubusercontent.com/assets/134455/25878072/9debbd1c-34ed-11e7-8930-aa85232564b9.png)

**Comment by whit537** [comment]
lol, duped existing aspen. :)

**Comment by whit537** [comment]
HEY @MATTBK COULD WE PLEASE LINK NPM PACKAGE TO EXISTING?!?!??! :P

**Comment by whit537** [comment]
Alright, we've landed #4305 and squished a few regressions. I find ten remaining tickets on the [project board](https://github.com/orgs/gratipay/projects/14?fullscreen=true). I've split them into "Heavy" (4) and "Light" (6) columns. I guess the reality is that we've got another three to six months of work here on this project.

**Comment by whit537** [comment]
I might prioritize #4476 over #4195. It should be simpler and it is something interesting for the giver side vs. the receiver side which is what we've been focusing on so far with #4305 and #4488. [slack](https://gratipay.slackarchive.io/gratipay/page-100/ts-1496784982968003)

**Comment by whit537** [comment]
Yeah, going for it. With #4476 we'll be able to minimally say that, "give to package.json." Everything else from there—and there's [plenty](https://github.com/orgs/gratipay/projects/14?fullscreen=true)!—is icing on the cake. Hope you like icing. 😁 🍰

**Comment by whit537** [comment]
#4476 landed in #4507. Now for some icing ... and bugfixes.

**Comment by whit537** [comment]
I folded the two Takeoff [columns](https://github.com/orgs/gratipay/projects/14?fullscreen=true) together.

**Comment by whit537** [comment]
So do we fix bugs and improve incrementally? Or do we try to get the whole dream workflow implemented barely with chicken wire first?

**Comment by whit537** [comment]
Incremental. Steady on, as we've been. #4305 took us four months. I started thinking through the Whole Dream Workflow and it's feeling like it will take about that again, three to six months as estimated https://github.com/gratipay/gratipay.com/issues/4427#issuecomment-302147796. Then we'll see if anyone wants it! 😅

**Comment by whit537** [comment]
I don't want to wait that long. 😞

**Comment by whit537** [comment]
Alright, well, I've kicked off a PR in #4569 for the Whole Dream Workflow. https://github.com/gratipay/gratipay.com/pull/4548 is a prerequisite for that as well as for #4539. Kick the can down the road! :-)

**Comment by whit537** [comment]
Turns out we need some significant improvements to our email infrastructure before we can move forward with the Whole Dream Workflow (WDW). Priorities I see there:

- Once we land the UI changes in #4579 #4584 that will mitigate the worst of our present phishing risk (#4557). There'll be more to do on that as we scale the WDW, but for now I think the next step is to ...
- Allow for sending emails to non-participants (#4548), which requires an unsubscribe/blacklist.
- Automated bounce/marked-as-spam handling (#2993) is close behind.
- Insta-send (#4545) is a nice-to-have but not critical for getting the WDW out the door initially.

**Comment by whit537** [comment]
Alright, I've kicked off #4598 to try to close the loop on giving to package.json. I am hoping we can move forward with a semi-manual process, because I don't want to spend six months building this out. It's so risky—we don't even know yet if anyone actually wants this enough to pay for it! Let's find out! :-)

In addition to the development work on #4598, we are going to need a ToS review, a HOWTO for handling the payments that come through #4598, and a blog post justifying a recommended dollar amount.

**Comment by whit537** [comment]
[slack](https://gratipay.slackarchive.io/gratipay/page-100/ts-1504618557000120)

**Comment by whit537** [comment]
Looks like we'll also need an admin dashboard for reviewing and distributing payments.

**Comment by whit537** [comment]
Might need an integration branch to bring together #4598 #4600 ... etc.?

**Comment by whit537** [comment]
Brag page could double as dashboard?

**Comment by whit537** [comment]
Blog post draft: https://gratipay.news/draft-your-company-should-probably-pay-draft-2-000-per-person-for-open-source-draft-9205443e209d

**Comment by whit537** [comment]
Got some feedback from @jdorfman on the blog post draft, and made some updates to it. 👍

**Comment by whit537** [comment]
> The definition of done is when we can say "Give us $10,000 and a package.json, and we'll distribute it for you."

Kind of amazing. This is getting stripped down, essentialized, bare-metaled in #4598:

> > The definition of done is when we can say "Give us $10,000 ~~and a package.json~~, and we'll distribute it for you."

**Comment by mattbk** [comment]
Not to add any more work tonight, but I mean they _could_ email us with suggestions in the interim before any uploading/parsing/ecosystems are added to the form.

**Comment by whit537** [comment]
This has evolved into https://github.com/gratipay/inside.gratipay.com/issues/1160. I'm shedding weight from this [project](https://github.com/orgs/gratipay/projects/14?fullscreen=true) so we can bring this in for a landing.
